packages/asset-swapper/src/network/chain.ts

```diff
@@ -0,0 +1,319 @@
+        }
+        if (!rawResultData || rawResultData === NULL_BYTES) {
+            rejectBatch(`EVM Dispatch call failed`);
+            return;
+        }
+        if (calls.length === 1) {
+            // Direct call (not batched).
+            return calls[0].accept(rawResultData);
+        }
+        const results = DISPATCHER_CONTRACT.getABIDecodedReturnData<DispatchedCallResult[]>('dispatch', rawResultData);
+        if (results.length !== calls.length) {
+            rejectBatch(`Expected dispatcher result to be the same length as number of calls`);
+            return;
+        }
+        // resolve each call in the batch.
+        results.forEach((r, i) => resolveQueuedCall(calls[i], r));
+    }
+
+    private static _mergeBatchCalls(calls: QueuedEthCall[]): BatchedChainEthCallOpts {
+        if (calls.length === 1) {
+            // If we just have one call, don't bother batching it.
+            return {
+                gas: calls[0].opts.gas || DEFAULT_CALL_GAS_LIMIT,
+                from: DEFAULT_CALLER_ADDRESS,
+                gasPrice: calls[0].opts.gasPrice || ZERO_AMOUNT,
+                to: calls[0].opts.to,
+                value: calls[0].opts.value || ZERO_AMOUNT,
+                data: calls[0].opts.data,
+                overrides: calls[0].opts.overrides || {},
+            };
+        }
+        const callInfos = calls.map(c =>
+            ({
+                data: c.opts.data,
+                to: c.opts.to,
+                gas: new BigNumber(c.opts.gas || 0),
+                value: c.opts.value || ZERO_AMOUNT,
+            }),
+        );
+        return {
+            gas: calls.reduce((s, c) => (c.opts.gas || DEFAULT_CALL_GAS_LIMIT) + s, 0) || DEFAULT_CALL_GAS_LIMIT,
+            from: DEFAULT_CALLER_ADDRESS,
+            gasPrice: BigNumber.sum(...calls.map(c => c.opts.gasPrice || 0)),
+            to: DISPATCHER_CONTRACT.address,
+            value: BigNumber.sum(...calls.map(c => c.opts.value || 0)),
+            data: DISPATCHER_CONTRACT.dispatch(callInfos).getABIEncodedTransactionData(),
+            overrides: Object.assign(
+                { [DISPATCHER_CONTRACT.address]: { code: DISPATCHER_CONTRACT_BYTECODE } },
+                ...calls.map(c => c.opts.overrides),
+            ),
+        };
+    }
+}
+
+function *generateCallBatches(queue: QueuedEthCall[]): Generator<QueuedEthCall[]> {
+    let nextQueue = queue;
+    while (nextQueue.length > 0) {
+        let _queue = nextQueue;
+        nextQueue = [];
+        const batch = [];
+        for (const c of _queue) {
+            if (canBatchCallWith(c, batch)) {
+                batch.push(c);
+            } else {
+                nextQueue.push(c);
+            }
+        }
+        yield batch;
+    }
+}
+
+function resolveQueuedCall(c: QueuedEthCall, r: DispatchedCallResult): void {
+    if (r.success) {
+        c.accept(r.resultData);
+    } else {
+        c.reject(tryDecodeStringRevertErrorResult(r.resultData) || `EVM call reverted`);
```

Comment:
Maybe reject it with the data?
```suggestion
        c.reject(tryDecodeStringRevertErrorResult(r.resultData) || `EVM call reverted: ${r.resultData}`);
```